audio/out/ao_sndio.c

```diff
@@ -0,0 +1,324 @@
+{
+    struct ao *ao = addr;
+    struct priv *p = ao->priv;
+
+    p->vol = newvol;
+}
+
+static int init(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+    struct mp_chmap_sel sel;
+    size_t i;
+    struct af_to_par {
+        int format, bits, sig;
+    };
+    static const struct af_to_par af_to_par[] = {
+        {AF_FORMAT_U8,   8, 0},
+        {AF_FORMAT_S16, 16, 1},
+        {AF_FORMAT_S32, 32, 1},
+    };
+    const struct af_to_par *ap;
+
+    p->hdl = sio_open(SIO_DEVANY, SIO_PLAY, 0);
+    if (p->hdl == NULL) {
+        MP_ERR(ao, "can't open sndio %s\n", SIO_DEVANY);
+        goto err_out;
+    }
+
+    sio_initpar(&p->par);
+    p->par.xrun = SIO_IGNORE;
+
+    /* Selecting sound format. */
+    ao->format = af_fmt_from_planar(ao->format);
+    for (i = 0, ap = af_to_par;; i++, ap++) {
+        if (i == MP_ARRAY_SIZE(af_to_par)) {
+            MP_VERBOSE(ao, "unsupported format\n");
+            p->par.bits = 16;
+            p->par.sig = 1;
+            p->par.le = SIO_LE_NATIVE;
+            break;
+        }
+        if (ap->format == ao->format) {
+            p->par.bits = ap->bits;
+            p->par.sig = ap->sig;
+            if (ap->bits > 8)
+                p->par.le = SIO_LE_NATIVE;
+            if (ap->bits != SIO_BPS(ap->bits))
+                p->par.bps = ap->bits / 8;
+            break;
+        }
+    }
+
+    p->par.rate = ao->samplerate;
+
+    /* Channels count. */
+    memset(&sel, 0x00, sizeof(sel));
+    for (i = 0; i < MP_ARRAY_SIZE(sndio_layouts); i++) {
+        mp_chmap_sel_add_map(&sel, &sndio_layouts[i]);
+    }
+    if (!ao_chmap_sel_adjust(ao, &sel, &ao->channels))
+        goto err_out;
+
+    p->par.pchan = ao->channels.num;
+    p->par.appbufsz = ((p->par.rate * 80) / 1000); /* 80ms buffer. */
+    p->par.round = ((p->par.rate * 10) / 1000); /* 10ms block size. */
+    if (!sio_setpar(p->hdl, &p->par)) {
+        MP_ERR(ao, "couldn't set params\n");
+        goto err_out;
+    }
+
+    /* Get current sound params. */
+    if (!sio_getpar(p->hdl, &p->par)) {
+        MP_ERR(ao, "couldn't get params\n");
+        goto err_out;
+    }
+    if (p->par.bps > 1 && p->par.le != SIO_LE_NATIVE) {
+        MP_ERR(ao, "swapped endian output not supported\n");
+        goto err_out;
+    }
+
+    /* Update sound params. */
+    if (p->par.bits == 8 && p->par.bps == 1 && !p->par.sig) {
+        ao->format = AF_FORMAT_U8;
+    } else if (p->par.bits == 16 && p->par.bps == 2 && p->par.sig) {
+        ao->format = AF_FORMAT_S16;
+    } else if ((p->par.bits == 32 || p->par.msb) && p->par.bps == 4 && p->par.sig) {
+        ao->format = AF_FORMAT_S32;
+    } else {
+        MP_ERR(ao, "couldn't set format\n");
+        goto err_out;
+    }
+
+    p->havevol = sio_onvol(p->hdl, volcb, ao);
+    sio_onmove(p->hdl, movecb, ao);
+
+    p->pfd = calloc(sio_nfds(p->hdl), sizeof(struct pollfd));
+    if (!p->pfd)
+        goto err_out;
+
+    ao->device_buffer = p->par.appbufsz;
```

Comment:
With ```p->par.bufsz``` does not work: GUI/Video freezes, looks like ```write()``` block, but sound - ok.
As I understand ```appbufsz``` - buf size available for unblocked write, and it can be set. ```bufsz``` - R/O internal allocated memory.

Reply:
I am testing your PR, but sound does not play for me, it just hangs at 0:00. This is because the buffer is never filled up to bufsz, so sndio never starts draining the buffer.

If I change line 301 (`state->free_samples`) to use `p->par.bufsz`, or if I change this line to use `p->par.appbufsz`, the issue goes away.

Reply:
```p->par.appbufsz = ((p->par.rate * 80) / 1000); /* 80ms buffer. */``` can you try increase it to 800?
```ao->device_buffer = p->par.bufsz / 2;``` - work for me, is ~82ms, so looks like it > appbufsz.

Reply:
If I set ```ao->device_buffer``` to ```p->par.appbufsz + p->par.round``` - 85,3ms still works.
I suspect that shdio wait to fill ```appbufsz``` and then starts, at least on OpenBSD, but FreeBSD wrapper starts after first ```sio_write()```.
```ao->device_buffer = p->par.appbufsz + p->par.round;``` - probably will allow both sndio implementations work.